lib/event_tracer/buffer.rb

```diff
@@ -58,8 +58,15 @@ def size
       attr_reader :buffer_size, :flush_interval, :buffer
 
       def add_item?
-        buffer.size < buffer_size &&
-          (buffer.empty? || buffer.first[:created_at] > Time.now - flush_interval)
+        return false if buffer.size >= buffer_size
+
+        # NOTE: we cannot use buffer.empty? then buffer.first here
+        # due to race-condition when another thread flushes the buffer
+        # right after buffer.empty? but before buffer.first is called.
+        # If we have more complicated use case, we may need to start introducing
+        # mutext.
+        first_item = buffer.first
+        first_item.nil? || first_item[:created_at] > Time.now - flush_interval
```

Comment:
is our buffer shared between threads sir?  I thought `Concurrent::Array` will lock the buffer for us to ensure no race condition 🤔

Reply:
> is our buffer shared between threads sir?

Yes. We initialize the buffer once in initializer, then use it for all puma/sidekiq threads. 

>  I thought Concurrent::Array will lock the buffer for us to ensure no race condition 🤔

No. `Concurrent::Array` doesn't help us in this case, as it's 2 separate calls. The class has no way to know that these two calls should be atomic.

Reply:
I see, thanks sir